developer-extension/src/panel/sessionReplayPlayer/startSessionReplayPlayer.ts

```diff
@@ -0,0 +1,167 @@
+) {
+  let status: SessionReplayPlayerStatus = 'loading'
+  const bufferedRecords = createRecordBuffer()
+
+  const messageBridge = createMessageBridge(iframe, () => {
+    const records = bufferedRecords.consume()
+    if (records.length > 0) {
+      status = 'ready'
+      onStatusChange(status)
+      records.forEach((record) => messageBridge.sendRecord(record))
+    } else {
+      status = 'waiting-for-full-snapshot'
+      onStatusChange(status)
+    }
+  })
+
+  const stopListeningToSdkMessages = listenSdkMessages((message) => {
+    if (message.type === 'record') {
+      const record = message.payload.record
+      if (status === 'loading') {
+        bufferedRecords.add(record)
+      } else if (status === 'waiting-for-full-snapshot') {
+        if (isFullSnapshotStart(record)) {
+          status = 'ready'
+          onStatusChange(status)
+          messageBridge.sendRecord(record)
+        }
+      } else {
+        messageBridge.sendRecord(record)
+      }
+    }
+  })
+
+  iframe.src = sandboxUrl
+
+  return {
+    stop() {
+      messageBridge.stop()
+      stopListeningToSdkMessages()
+    },
+  }
+}
+
+function createRecordBuffer() {
+  const records: BrowserRecord[] = []
+
+  return {
+    add(record: BrowserRecord) {
+      // Make sure 'records' starts with a FullSnapshot
+      if (isFullSnapshotStart(record)) {
+        records.length = 0
+        records.push(record)
+      } else if (records.length > 0) {
+        records.push(record)
+      }
+    },
+    consume(): BrowserRecord[] {
+      return records.splice(0, records.length)
+    },
+  }
+}
+
+function isFullSnapshotStart(record: BrowserRecord) {
+  // All FullSnapshot start with a "Meta" record. The FullSnapshot record comes in third position
+  return record.type === RecordType.Meta
+}
+
+function normalizeRecord(record: BrowserRecord) {
+  if (record.type === RecordType.IncrementalSnapshot && record.data.source === IncrementalSource.MouseMove) {
+    return {
+      ...record,
+      data: {
+        ...record.data,
+        position: record.data.positions[0],
+      },
+    }
+  }
+  return record
+}
+
+function createMessageBridge(iframe: HTMLIFrameElement, onReady: () => void) {
+  let nextMessageOrderId = 1
+
+  function globalMessageListener(event: MessageEvent<MessageBridgeUp>) {
+    if (event.origin === sandboxOrigin) {
+      const message = event.data
+      if (message.type === 'log') {
+        if (message.level === 'error') {
+          sandboxLogger.error(message.message)
+        } else {
+          sandboxLogger.log(message.message)
+        }
+      } else if (message.type === 'error') {
+        sandboxLogger.error(
+          `${message.serialisedError.name}: ${message.serialisedError.message}`,
+          message.serialisedError.stack
+        )
+      } else if (message.type === 'ready') {
+        onReady()
+      } else {
+        // sandboxLogger.log('Unhandled message type:', message)
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏should we keep this commented log?